### PR TITLE
Add V2 filtering logic that detects exposures given an array of `ExposureWindow`s

### PR DESCRIPTION
**Pull request by smcmurtry**
# Summary | Résumé

This PR adds V2 filtering logic that detects exposures given an array of `ExposureWindow`s. I have added:
- `checkIfExposedV2()`: this takes an array of `ExposureWindow`s, bluetooth attenuation thresholds, and the 15 minute minimum and checks if the provided array contains an exposure
- `performExposureStatusUpdateV2()` is meant to be interchangeable with the existing function `performExposureStatusUpdate()`, but it uses the new V2 functionality

These functions are not called anywhere yet and should not be until we have implemented `getExposureWindows()` and `provideDiagnosisKeys()` in the native layer. I am testing `checkIfExposedV2()` w/ mock data.

This is one approach - an alternative would be to use the new `getDailySummaries()` function.

[Google V2 API docs](https://developers.google.com/android/exposure-notifications/exposure-notifications-api).

### CODE REVIEW

**Comment by smcmurtry** [comment]
Posting some review comments from others:
- split V2 tests into a separate file
- pass args to `checkIfExposedV2` as an object for readability
- split the scan instance logic in the for loop into its own fn
- populate a new summary obj to save in the status, or come up w/ an alternative (may need to calc the last exposed timestamp`)
- sort exposure windows in order of most recent to help w/ the above

**Comment by smcmurtry** [comment]
A few more todos:
- [ ] save a "fake" summary obj to the status in the V2 code, this will require:
- [ ] calculate `matchedKeyCount`
- [ ] calculate `lastExposureTimestamp`
- [ ] calculate `attenuationDurations`
- [ ] write test to check `matchedKeyCount` is correct
- [ ] write test to check `lastExposureTimestamp` is correct
- [ ] write tests for "multiple exposures in the same day" scenario

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 1]
```diff
@@ -341,6 +344,120 @@ export class ExposureNotificationService {
...
+      .map(x => x.secondsSinceLastScan)
+      // with initial value to avoid when the array is empty
+      .reduce((partialSum, x) => partialSum + x, 0);
+    return totalSeconds;
+  };
+
+  public getAttenuationDurations = (scanInstances: ScanInstance[], attenuationDurationThresholds: number[]) => {
+    //  typicalAttenuation values are in positive DB units
+    const immediateScans = scanInstances.filter(x => x.typicalAttenuation <= attenuationDurationThresholds[0]);
+    const nearScans = scanInstances.filter(x => {
```
Comment (jeberhardt) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:358]: Please change the variable "x" to something more meaningful.

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 2]
```diff
@@ -341,6 +344,120 @@ export class ExposureNotificationService {
...
+    attenuationDurationThresholds: number[];
+    minimumExposureDurationMinutes: number;
+  }): Promise<[boolean, ExposureSummary | undefined]> {
+    if (exposureWindows.length === 0) {
+      return [false, undefined];
+    }
+    const dailySummariesObj: {[key: string]: {attenuationDurations: number[]; matchedKeyCount: number}} = {};
+    exposureWindows
+      .filter(window => {
+        if (window.infectiousness === Infectiousness.None) {
```
Comment (jeberhardt) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:383]: This `if` statement can be simplified, combined with the `return` statement.

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 3]
```diff
@@ -341,6 +344,120 @@ export class ExposureNotificationService {
...
+      .forEach(windowSummary => {
+        const dayString = windowSummary.day.toString();
+        dailySummariesObj[dayString].attenuationDurations[0] += windowSummary.attenuationDurations[0];
+        dailySummariesObj[dayString].attenuationDurations[1] += windowSummary.attenuationDurations[1];
+        dailySummariesObj[dayString].attenuationDurations[2] += windowSummary.attenuationDurations[2];
+        dailySummariesObj[dayString].matchedKeyCount += 1;
+      });
+
+    const dailySummaries = Object.entries(dailySummariesObj)
+      .map(x => {
```
Comment (jeberhardt) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:404]: Please change the variable "x" to something more meaningful.

File: src/services/ExposureNotificationService/ExposureNotificationService.ts [hunk 4]
```diff
@@ -341,6 +344,120 @@ export class ExposureNotificationService {
...
+          {
+            type: ExposureStatusType.Exposed,
+            summary: dailySummary,
+          },
+          lastCheckedPeriod,
+        );
+      }
+      return this.finalize({}, lastCheckedPeriod);
+    } catch (error) {
+      captureException('performExposureStatusUpdateV2', error);
```
Comment (jeberhardt) [inline src/services/ExposureNotificationService/ExposureNotificationService.ts:455]: Should this `return false` to reject the promise?

**Review by jeberhardt — CHANGES_REQUESTED** [review]

File: src/services/ExposureNotificationService/ExposureNotificationService.ts (hunk 2 above)
Comment (smcmurtry) [reply src/services/ExposureNotificationService/ExposureNotificationService.ts:383]: I find my version easier to read but I will go with your suggestion anyway 😉

File: src/services/ExposureNotificationService/ExposureNotificationService.ts (hunk 4 above)
Comment (jeberhardt) [reply src/services/ExposureNotificationService/ExposureNotificationService.ts:452]: Shouldn't it return a boolean value, rather than just void?

**Review by jeberhardt — APPROVED** [review]

**Pull request merged**